FEATURE/14 JWT 로그인 기능 구현

jieun5119 opened the pull request:
<!--
  템플릿은 아직 PR 작성이 익숙하지 않으신 분들을 위해서 제공하는 가이드입니다!
  리뷰어 또는 이 PR을 보게 될 다른 사람들이 이 PR을 보는데 참고할 수 있는 내용이 있다면 포함해서 작성해주시면 됩니다.
-->
resolved : #14 
## 📌 과제 설명 
JWT 기반 로그인 기능을 구현하였고, 리프레시 토큰을 통한 토큰 재발급과 스케줄링 기능도 추가하였습니다. 

## 👩‍💻 요구 사항과 구현 내용
- jwt의존성 추가 및 security 설정
- 로그인 필터 구현 및 등록
- jwt 필터 등록
- 리프레시 토큰 발급
- 리프레시 토큰 스케줄링 기능 
-
## ✅ 피드백 반영사항  

- 코드 중복 제거를 위해 함수화 진행
- 검증 로직 별도의 메서드로 분
-  NPE(NullPointerException) 방지 처리 추가
- `ReissueService`에서 예외 처리(void로 수정)

## ✅ PR 포인트 & 궁금한 점 
- JWT 기반 로그인과 리프레시 토큰 관리 부분에서 추가적인 피드백이 있는지 궁금합니다.
- 레디스 대신 저는 RDB에 저장하고 토큰이 쌓이면 삭제 해주기 위해 스케줄링을 구현했는데 이 부분도 어떻게 생각하시는지 궁금합니다.

jieun5119 committed:
jwt의존성 추가 및 Security 설정 JWT 방식으로 변경

jieun5119 committed:
jwt의존성 추가 및 Security 설정 JWT 방식으로 변경

jieun5119 committed:
feat: 로그인 필터 구현 및 등록

- 로그인 요청 및 검증 기본 구현

- 로그인 필터 등록

jieun5119 committed:
feat: 로그인 필터 구현 및 등록

- 로그인 요청 및 검증 기본 구현

- 로그인 필터 등록

jieun5119 committed:
feat: JWT 필터 및 로그인 필터 구현

- JWT 필터에서 access 토큰 만료 및 카테고리 검증 로직 추가

- 로그인 성공 시 access, refresh 토큰 생성 및 refresh 토큰 쿠키 저장

- 인증 실패 시 예외 처리 로직 수정

jieun5119 committed:
feat: JWT 필터 및 로그인 필터 구현

- JWT 필터에서 access 토큰 만료 및 카테고리 검증 로직 추가

- 로그인 성공 시 access, refresh 토큰 생성 및 refresh 토큰 쿠키 저장

- 인증 실패 시 예외 처리 로직 수정

jieun5119 committed:
feat: 리프레시 토큰 관리 기능 구현

- 리프레시 토큰 테이블 엔티티 및 레포지토리 구성

- 토큰 재발급 서비스 로직 추가: 기존 리프레시 토큰 삭제 후 새로운 토큰 발급 및 쿠키 저장

- 만료된 리프레시 토큰 처리 및 예외 처리 로직 추가

jieun5119 committed:
feat: 리프레시 토큰 관리 기능 구현

- 리프레시 토큰 테이블 엔티티 및 레포지토리 구성

- 토큰 재발급 서비스 로직 추가: 기존 리프레시 토큰 삭제 후 새로운 토큰 발급 및 쿠키 저장

- 만료된 리프레시 토큰 처리 및 예외 처리 로직 추가

jieun5119 committed:
fix: 관련 DTO 추가 누락

- 리프레시 토큰 , UserDetails 관련 DTO를 추가

jieun5119 committed:
fix: 관련 DTO 추가 누락

- 리프레시 토큰 , UserDetails 관련 DTO를 추가

jieun5119 committed:
feat: 리프레시 토큰 스케줄링 기능 추가

- 자정마다 만료된 리프레시 토큰을 삭제하는 스케줄러 추가

- RefreshRepository에 만료된 토큰 삭제 메서드 추가

jieun5119 committed:
feat: 리프레시 토큰 스케줄링 기능 추가

- 자정마다 만료된 리프레시 토큰을 삭제하는 스케줄러 추가

- RefreshRepository에 만료된 토큰 삭제 메서드 추가

jieun5119 committed:
불필요한 로그 파일 삭제 및 저장소 정리

jieun5119 committed:
불필요한 로그 파일 삭제 및 저장소 정리

jieun5119 committed:
refactor: @Transactional 레포지토리에서 서비스로 이동

jieun5119 committed:
refactor: @Transactional 레포지토리에서 서비스로 이동

jieun5119 committed:
fix: @Transactional 누락 추가

jieun5119 committed:
fix: 예외처리 수정

Leehunil commented on src/main/java/com/sscanner/team/jwt/JWTUtil.java:
```diff
@@ -0,0 +1,54 @@
+        this.secretKey = new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), Jwts.SIG.HS256.key().build().getAlgorithm());
+    }
+
+    public String getCategory(String token){
```
Jwts.parser().verifyWith(secretKey).build().parseSignedClaims(token).getPayload() 이 부분을 함수화 시키면 코드 중복을 줄일 수 있을 것 같습니다!!

jieun5119 replied:
넵 코드 중복 부분 함수화 시켜 수정했습니다 !

Leehunil commented on src/main/java/com/sscanner/team/user/service/ReissueService.java:
```diff
@@ -0,0 +1,93 @@
+    private final RefreshRepository refreshRepository;
+
+    public ApiResponse<RefreshResponseDto> reissueToken(HttpServletRequest request, HttpServletResponse response) {
+        String refresh = null;
```
이 부분은 로직이 너무 길어서 가독성이 떨이지는 것 같아요~ 조금 세분화 시키면 좋을 것 같아요!!
예를 들면 public void checkRefreshTokenExist() 이렇게 세분화 하면 더 가독성이 좋아질 것 같아요

l2yujw replied:
훈일님 말씀대로 세분화 하시는거 좋은거 같습니다!
추가적으로 refresh를 반환하게 될테니까 return cookie.getValue()로 하고
null 체크대신 return을 못했을 때 throw new ~ 를 바로 던지는 것도 괜찮지 않을까 생각했습니다!

jieun5119 replied:
이 부분도 가독성을 고려해 세분화 진행했습니다.
정원님이 말하신 부분도 참고해서 예외 처리 하도록 수정했습니다

Leehunil reviewed (commented):
제가 해석하기로는 UsernamePasswordAuthenticationFilter를 커스텀해서 RDB에 토큰 저장하고 쿠키를 생성해서 안에 토큰을 넣어서 관리하는 것 같아요 맞나요...?? 처음 보는 방식이라 흥미로웠습니다!! 쿠키에 토큰을 넣어서 관리하면 어떠한 장점이 있는지 궁금해요~

7zrv commented on src/main/java/com/sscanner/team/jwt/LoginFilter.java:
```diff
@@ -0,0 +1,95 @@
+        refreshRepository.save(refreshEntity);
+    }
+
+    private Cookie createCookie(String key, String value) {
```
쿠키 방식을 선택하신 이유가 궁금해요!

jieun5119 replied:
제가 refresh 토큰을 쿠키에 저장한 이유는 슬랙에 정리해뒀습니다!  읽어보시고 다른 의견이 있으시면 피드백 부탁드릴게요

l2yujw commented on src/main/java/com/sscanner/team/user/service/ReissueService.java:
```diff
@@ -0,0 +1,93 @@
+        if (refresh == null) {
+            throw new BadRequestException(ExceptionCode.NOT_EXIST_REFRESH_TOKEN);
+        }
+
```
validateXXX(String refresh) 메서드로 검증 로직을 따로 빼는건 어떠신가요?!

jieun5119 replied:
넵 따로 메서드를 구현하는 방향으로 수정했습니다

l2yujw commented on src/main/java/com/sscanner/team/user/service/ReissueService.java:
```diff
@@ -0,0 +1,93 @@
+
+        // 리프레시 토큰이 맞는지 체크
+        String category = jwtUtil.getCategory(refresh);
+        if (!category.equals("refresh")) {
```
"refresh".equals(category) 이렇게 작성하면 NPE 방지가 가능합니다

l2yujw commented on src/main/java/com/sscanner/team/jwt/JWTUtil.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public Boolean isExpired(String token) {
+        return Jwts.parser().verifyWith(secretKey).build().parseSignedClaims(token).getPayload().getExpiration().before(new Date());
```
ReissueService에서 ExpiredJwtException 을 잡는걸로 봤는데 ExpiredJwtException이 자동으로 던져지는건가요?!
자동으로 던져지는거면 Boolean대신 void를 써도 될 것 같습니다!

jieun5119 replied:
넵 내부 라이브러리에서 자동으로 처리되는 부분이기 때문에 정원님이 말씀하신대로 void로 수정 완료했습니다

jieun5119 committed:
fix: 로그인 기능 피드백 반영

- 함수화를 통한 코드 중복 제거(파서)

- 검증 로직 별도의 메서드로 분리

- NPE방지 처리 추가

ReissueService에서 예외처리 수정(void로 수정)

jieun5119 committed:
fix:isExpired메소드 수정

- 불필요한 주석 제거

parseToken사용

jieun5119 committed:
fix:불필요한 에러코드 제거

jieun5119 committed:
Merge branch 'develop' into 14-add-user-login

jieun5119 committed:
fix: 예외코드 추가

jieun5119 committed:
fix: 상수 적용

- refresh, access 상수 적용

jieun5119 committed:
fix : 메소드명, 필드명 수정
